samples/subsys/usb/hid/src/main.c

```diff
@@ -11,123 +11,137 @@
-	HID_REPORT_SIZE(8),
-	/* 0x95, 0x01,			REPORT_COUNT (1)		*/
-	HID_REPORT_COUNT(1),
-	/* 0x09, 0x00,			USAGE (Undefined)		*/
-	HID_USAGE(HID_USAGE_GEN_DESKTOP_UNDEFINED),
-	/* 0x91, 0x82,			OUTPUT (Data,Var,Abs,Vol)	*/
-	HID_OUTPUT(0x82),
-	/* 0xc0			END_COLLECTION			*/
 	HID_END_COLLECTION,
 };
 
 static void send_report(struct k_work *work)
 {
-	static uint8_t report_1[2] = { REPORT_ID_1, 0x00 };
 	int ret, wrote;
 
-	ret = hid_int_ep_write(hdev, report_1, sizeof(report_1), &wrote);
-
-	LOG_DBG("Wrote %d bytes with ret %d", wrote, ret);
-
-	/* Increment reported data */
-	report_1[1]++;
+	if (!atomic_test_and_set_bit(hid_ep_in_busy, HID_EP_BUSY_FLAG)) {
+		ret = hid_int_ep_write(hdev, (uint8_t *)&report_1,
+				       sizeof(report_1), &wrote);
+		if (ret != 0) {
+			LOG_ERR("Failed to submit report");
+		} else {
+			LOG_DBG("Report submitted");
+		}
+	} else {
+		LOG_DBG("HID IN endpoint busy");
+	}
 }
 
-static void in_ready_cb(const struct device *dev)
+static void int_in_ready_cb(const struct device *dev)
 {
 	ARG_UNUSED(dev);
-
-	k_delayed_work_submit(&delayed_report_send, REPORT_TIMEOUT);
+	if (!atomic_test_and_clear_bit(hid_ep_in_busy, HID_EP_BUSY_FLAG)) {
+		LOG_WRN("IN endpoint callback without preceding buffer write");
+	}
 }
 
-static void status_cb(enum usb_dc_status_code status, const uint8_t *param)
+/*
+ * On Idle callback is available here as an example even if actual use is
+ * very limited. In contrast to report_event_handler(),
+ * report value is not incremented here.
+ */
+static void on_idle_cb(const struct device *dev, uint16_t report_id)
 {
-	switch (status) {
-	case USB_DC_CONFIGURED:
-		in_ready_cb(hdev);
-		break;
-	case USB_DC_SOF:
-		break;
-	default:
-		LOG_DBG("status %u unhandled", status);
-		break;
-	}
+	LOG_DBG("On idle callback");
+	k_work_submit(&report_send);
 }
 
-static void idle_cb(const struct device *dev, uint16_t report_id)
+static void report_event_handler(struct k_timer *dummy)
 {
-	static uint8_t report_1[2] = { 0x00, 0xEB };
-	int ret, wrote;
-
-	ret = hid_int_ep_write(dev, report_1, sizeof(report_1), &wrote);
-
-	LOG_DBG("Idle callback: wrote %d bytes with ret %d", wrote, ret);
+	/* Increment reported data */
+	report_1.value++;
+	k_work_submit(&report_send);
 }
 
+K_TIMER_DEFINE(event_timer, report_event_handler, NULL);
+
 static void protocol_cb(const struct device *dev, uint8_t protocol)
 {
-	LOG_DBG("New protocol: %s", protocol == HID_PROTOCOL_BOOT ?
+	LOG_INF("New protocol: %s", protocol == HID_PROTOCOL_BOOT ?
 		"boot" : "report");
 }
 
 static const struct hid_ops ops = {
-	.int_in_ready = in_ready_cb,
-	.on_idle = idle_cb,
+	.int_in_ready = int_in_ready_cb,
+	.on_idle = on_idle_cb,
 	.protocol_change = protocol_cb,
 };
 
+static void status_cb(enum usb_dc_status_code status, const uint8_t *param)
+{
+	switch (status) {
+	case USB_DC_RESET:
+		configured = false;
+		break;
+	case USB_DC_CONFIGURED:
+		if (!configured) {
+			int_in_ready_cb(hdev);
```

Comment:
Perhaps it would be better to only clear `hid_ep_in_busy` here, as this is the intention of this call, right?

Reply:
It is roughly consistent with other classes and I would rather keep it that way.